### PR TITLE
Automate architecture synchronization with latest LLVM

### PR DESCRIPTION
**Pull request by Phosphorus15**
## What is Auto-Sync ?

Capstone Auto-Sync is an initiative to partly automate the synchronization of certain architectures to the latest.

Most of the Capstone's `.inc` files a generated from LLVM's TableGen backend and processed by python scrips in `suite/synctools` into C-compatible files, which leads to the problem that with LLVM's update, it's not always (hardly, in fact) possible to use the synctools without patch in regard to LLVM's upstream change.

This syncing tools, however, using a custom-made LLVM TableGen backend ([here](https://github.com/rizinorg/llvm-capstone)) to generate `.inc` files natively usable by Capstone. With certain adaptations in Capstone's structure, it is possible to consistently automate large parts of the work on keeping up with LLVM's latest Target (i.e. `.td` files) update, and optimally, there could be zero-overhead in the process of updating(see [this patch on missing bcxf instructions](https://github.com/rizinorg/llvm-capstone/commit/d594f4f6b0755ab11580e1e87d610b560e71b5ef))

https://github.com/Phosphorus15/Capstone/blob/1bca4211bd7c132c57d0006a26de57d42a4bcdb9/sync/SYNCING.md?plain=1#L1-L17

## What is done in this PR?

In this PR, we adapted following architectures with the Auto-Sync structure base on the `next` branch, and made sure they pass the test suite given by Capstone, and also by the analysis tools [rizin](https://github.com/rizinorg/rizin) (which also uses Capstone):

- Mips
- ARM
- AArch64
- Riscv
- PowerPC
- Sparc
- SystemZ
- XCore

## What is yet to be done?

Capstone's instructions mapping info (like enums) exposed to various bindings (Python bindings, etc.) are not affected by Auto-Sync procedure, and so have to be manually edited if anything like new instructions/operands type was added by LLVM.

### CODE REVIEW

**Comment by Phosphorus15** [comment]
It is also recommended to look into https://github.com/Phosphorus15/capstone-1/blob/auto-sync-integ/sync/SYNCING.md and https://github.com/Phosphorus15/capstone-1/blob/auto-sync-integ/sync/README.md for more information

**Comment by aquynh** [comment]
Very cool! Did you try this on the latest LLVM version, and was anything
broken?

**Comment by Phosphorus15** [comment]
> Very cool! Did you try this on the latest LLVM version, and was anything broken?

Actually yes, the generated .inc files were based on the latest commit of (llvm-project)[https://github.com/llvm/llvm-project], and we tested it against cstest so it works very well on most test cases, **except** that LLVM drops PowerPC QPX support (so this table could not be gen-ed), and a couple of thumb2 instructions are missing.

**Comment by aquynh** [comment]
Can you provide some quick summary on what changed for x86, arm & arm64
with this PR? This brings latest instruction update to all of them?

**Comment by aquynh** [comment]
did you base this PR on some development snapshot of llvm?

since development snapshot of llvm may include some bugs or WIP, I think it is better to base on stable version, which is llvm 13.0 at the moment.  i am just wondering if that is a lot of work to PR based that version now?

File: cs.c [hunk 1]
```diff
@@ -891,8 +896,9 @@ size_t CAPSTONE_API cs_disasm(csh ud, const uint8_t *buffer, size_t size, uint64
 		mci.flat_insn->mnemonic[0] = '\0';
 		mci.flat_insn->op_str[0] = '\0';
 #endif
-
+        debugln("trying into dissasm");
```
Comment (aquynh) [inline cs.c:899]: can you remove the debug code?

**Comment by aquynh** [comment]
please move sync/ directory to suite/

**Comment by XVilka** [comment]
@Phosphorus15 could you please address this warning too?
```
In file included from /src/capstonenext/arch/AArch64/AArch64InstPrinter.c:225:
/src/capstonenext/arch/AArch64/AArch64GenDisassemblerTables.inc:157863:19: warning: passing 'const MCInst *' (aka 'const struct MCInst *') to parameter of type 'MCInst *' (aka 'struct MCInst *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    printZPRasFPR(MI, OpIdx, OS, 128);
                  ^~
/src/capstonenext/arch/AArch64/AArch64InstPrinter.c:187:35: note: passing argument to parameter 'MI' here
static void printZPRasFPR(MCInst *MI, unsigned OpNum, SStream *O, int Width);
                                  ^
In file included from /src/capstonenext/arch/AArch64/AArch64InstPrinter.c:225:
/src/capstonenext/arch/AArch64/AArch64GenDisassemblerTables.inc:157867:19: warning: passing 'const MCInst *' (aka 'const struct MCInst *') to parameter of type 'MCInst *' (aka 'struct MCInst *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    printZPRasFPR(MI, OpIdx, OS, 32);
                  ^~
/src/capstonenext/arch/AArch64/AArch64InstPrinter.c:187:35: note: passing argument to parameter 'MI' here
static void printZPRasFPR(MCInst *MI, unsigned OpNum, SStream *O, int Width);
                                  ^
In file included from /src/capstonenext/arch/AArch64/AArch64InstPrinter.c:225:
/src/capstonenext/arch/AArch64/AArch64GenDisassemblerTables.inc:157871:27: warning: passing 'const MCInst *' (aka 'const struct MCInst *') to parameter of type 'MCInst *' (aka 'struct MCInst *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    printMatrixTileVector(MI, OpIdx, OS, 0);
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -29,7 +29,7 @@ option(CAPSTONE_BUILD_CSTOOL "Build cstool" ON)
 option(CAPSTONE_USE_DEFAULT_ALLOC "Use default memory allocation functions" ON)
 option(CAPSTONE_ARCHITECTURE_DEFAULT "Whether architectures are enabled by default" ON)
 option(CAPSTONE_DEBUG "Whether to enable extra debug assertions" OFF)
-option(CAPSTONE_INSTALL "Generate install target" OFF)
```
Comment (mrexodia) [inline CMakeLists.txt:32]: This was done like this on purpose to support `FetchContent` workflows without doing custom hackery on the cache.

**Comment by Phosphorus15** [comment]
> did you base this PR on some development snapshot of llvm?
> 
> since development snapshot of llvm may include some bugs or WIP, I think it is better to base on stable version, which is llvm 13.0 at the moment. i am just wondering if that is a lot of work to PR based that version now?

This surely could be (by rebasing the tablegen backend onto llvm 13.0) done with no much of a problem, just that we'd lose the update of some new instructions on ARMv9 in the lastest dev branch of LLVM.

**Comment by Phosphorus15** [comment]
> Can you provide some quick summary on what changed for x86, arm & arm64 with this PR? This brings latest instruction update to all of them?

x86 has not been included because it has a quite different tablegen backend implementation than other archs., also what is the `change` specifically, instructions change or changes on Capstone code structure? I can write a small documentation about it.

**Comment by aquynh** [comment]
That is fine to miss something in the latest version, since we will always
miss the latest update. Even this PR is out of date now since yesterday, if
you think about the quick progress of the dev branch of llvm.

**Comment by aquynh** [comment]
About both, please.

File: CMakeLists.txt (hunk 2 above)
Comment (thestr4ng3r) [reply CMakeLists.txt:32]: Regardless, the change seems unrelated to the goal of this pr, so it should not be part of it.

**Comment by XVilka** [comment]
It also should be rebased since there are now conflicts, @Phosphorus15

**Comment by kabeor** [comment]
@XVilka emm..PR to 'next' branch is right, why need to rebase?

**Comment by XVilka** [comment]
@kabeor because there were recent commits in the `next`, and there is a conflict in `arch/AArch64/AArch64InstPrinter.c`: https://github.com/capstone-engine/capstone/commits/next

**Comment by kabeor** [comment]
@XVilka we will solve the conflict. 'next' branch is main branch for now.

**Comment by kabeor** [comment]
@Phosphorus15 Hello, is there any progress? or some thing I can help?

**Comment by XVilka** [comment]
@Phosphorus15 ping?

**Comment by FlyGoat** [comment]
I'm trying to introduce MIPS R6 support and just found this PR had done that :-)

In case the original author stalled, is it possible to get this merged by maintainer edit?

Or I can try to rebase and address these review comments, but how to credit the author properly?

**Comment by XVilka** [comment]
@FlyGoat I think you can do a new PR, it would be the best way to handle this. You can just keep original copyrights and add yours. You also could link to the original PR. Since the patched LLVM resides in our repository, feel free to send PRs here as well: https://github.com/rizinorg/llvm-capstone

**Comment by FlyGoat** [comment]
@XVilka Thanks for the information. I realized that a huge amount of diffs in this patch is doing `clang-format` so I would like to get a confirmation from maintainers before I mess with them.
@aquynh 
Is formatting necessary, or welcomed?

**Comment by kabeor** [comment]
@FlyGoat Hello, we must keep the original code style, so we won't accept code formatting changes in this PR.

In this PR, what we need to do are:

1. create a new PR rebase to new branch based on next branch, we're considering merging it until it's stable.
2. restore all code formatting changes
3. do not add any .inc file in this PR, we should PR that alone
4. the code should based on LLVM latest released version, but not latest commit(because there could be more bugs there)

@XVilka For more convenient maintenance in the future, we really hope to move llvm-capstone repo to Capstone Organization. Maybe we can talk more about this.

**Comment by XVilka** [comment]
@kabeor yes, of course, that was the original intention once the PR is in the better shape to get merged.

**Comment by kabeor** [comment]
> @kabeor yes, of course, that was the original intention once the PR is in the better shape to get merged.

Cool!

**Comment by XVilka** [comment]
@kabeor I transferred the repository to you since cannot transfer to the capstone-engine directly.

**Comment by kabeor** [comment]
@XVilka ok, I'll transfer now, thx

**Comment by kabeor** [comment]
this here now https://github.com/capstone-engine/llvm-capstone

**Comment by kabeor** [comment]
@XVilka hi, Would you like to maintain llvm-capstone continuing with us together? I can give you guys the access.

**Comment by FlyGoat** [comment]
For `llvm-capestone` repo I guess what we need to do is squash TableGen backend into a commit and rebase to llvm-13 stable branch?

**Comment by kabeor** [comment]
@FlyGoat yes

**Comment by XVilka** [comment]
@kabeor yes, please do.

**Comment by kabeor** [comment]
> @kabeor yes, please do.

done ：)

**Comment by Rot127** [comment]
I'm would take this PR and https://github.com/capstone-engine/capstone/pull/1831 over and finish them up. If anyone already started on this as well in the last last 5 months please ping me here.
The plan is to extend it also by the LANAI and LoongArich. But of cause the modernization of this PR comes first.

**Comment by XVilka** [comment]
Probably could be closed in favor of https://github.com/capstone-engine/capstone/pull/1949 
@kabeor

**Pull request closed without merging**